### PR TITLE
feat: smart account refactor

### PR DESCRIPTION
**Pull request by lukaisailovic**
Current implementation does not follow previously established principles in the repo. Refactoring smart account implementation to support multiple account types (Kernel,Safe,Biconomy...) and optimize to not deploy on session request.

Additions:
- Add support for Kernel Smart Account
- Parsing getPermissions from signTypedData and returning session keys

### CODE REVIEW

**Review by svenvoskamp — APPROVED** [review]

File: advanced/wallets/react-wallet-v2/package.json [hunk 1]
```diff
@@ -31,14 +31,19 @@
     "@taquito/signer": "^15.1.0",
     "@taquito/taquito": "^15.1.0",
     "@walletconnect/web3wallet": "1.10.2",
+    "@zerodev/ecdsa-validator": "^5.1.0",
+    "@zerodev/presets": "^5.1.4",
+    "@zerodev/sdk": "^5.1.8",
+    "@zerodev/session-key": "^5.1.1",
+    "@zerodev/weighted-ecdsa-validator": "^5.1.0",
```
Comment (tomiir) [inline advanced/wallets/react-wallet-v2/package.json:38]: we should not use ^ versions

File: advanced/wallets/react-wallet-v2/src/components/PermissionDetailsCard.tsx [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+  description: string
+}
+
+interface IProps {
+  scope: PermissionAction[]
+}
+
+export default function PermissionDetailsCard({ scope }: IProps) {
+  return (
+    <>
```
Comment (tomiir) [inline advanced/wallets/react-wallet-v2/src/components/PermissionDetailsCard.tsx:13]: nit: outermost Fragment not needed

File: advanced/wallets/react-wallet-v2/src/hooks/useInitialization.ts [hunk 3]
```diff
@@ -29,6 +30,14 @@ export default function useInitialization() {
       const { tronAddresses } = await createOrRestoreTronWallet()
       const { tezosAddresses } = await createOrRestoreTezosWallet()
       const { kadenaAddresses } = await createOrRestoreKadenaWallet()
+      
+
+      if(smartAccountEnabled){
+        if(kernelSmartAccountEnabled){
```
Comment (tomiir) [inline advanced/wallets/react-wallet-v2/src/hooks/useInitialization.ts:36]: wouldn't this be an `&&` ?

File: advanced/wallets/react-wallet-v2/src/hooks/usePriorityAccounts.ts [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+
+    if (smartAccountEnabled) {
+        /**
+         * If our kernel account supports any of the requested chainIds, put it in the first spot
+         */
+        if (kernelSmartAccountEnabled) {
+          return kernelAddressPriority(namespaces,kernelSmartAccountAddress)
+        }
+    }
+    return []
```
Comment (tomiir) [inline advanced/wallets/react-wallet-v2/src/hooks/usePriorityAccounts.ts:22]: what happens if SA is enabled but kernel isn't? is that possible?

File: advanced/wallets/react-wallet-v2/src/lib/KernelSmartAccountLib.ts [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+    private validator: KernelValidator | undefined
+
+
+    #signerPrivateKey: string;
+
+    public constructor({ privateKey, chain, sponsored = false }: SmartAccountLibOptions) {        
+        this.chain = chain
+        this.sponsored = sponsored
+        this.#signerPrivateKey = privateKey
+        this.signer = privateKeyToAccount(privateKey as Hex)
```
Comment (tomiir) [inline advanced/wallets/react-wallet-v2/src/lib/KernelSmartAccountLib.ts:43]: for future: if Kernel allows signers, a couple of static methods to construct from signers or PK would be nice

File: advanced/wallets/react-wallet-v2/src/lib/KernelSmartAccountLib.ts [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+
+        this.validator = await createWeightedECDSAValidator(this.publicClient, {
+            config: {
+                threshold: 100,
+                signers: [
+                  { address: this.signer.address, weight: 100 },
+                ]
+              },
+              signers: [this.signer],
+        })
```
Comment (tomiir) [inline advanced/wallets/react-wallet-v2/src/lib/KernelSmartAccountLib.ts:66]: this is lit 🔥

File: advanced/wallets/react-wallet-v2/src/lib/KernelSmartAccountLib.ts [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+        if (!this.publicClient) {
+            throw new Error("Client not initialized");
+        }
+        if(!address){
+            throw new Error("Target address is required");
+        }
+        const parsedPermissions = JSON.parse(permissions)
+        const sessionKeyAddress = address
+        console.log('Issuing new session key',{sessionKeyAddress});
+        const emptySessionKeySigner = addressToEmptyAccount(sessionKeyAddress)
```
Comment (tomiir) [inline advanced/wallets/react-wallet-v2/src/lib/KernelSmartAccountLib.ts:166]: curious: what's an `emptySmartAcconut`  ?

File: advanced/wallets/react-wallet-v2/src/pages/settings.tsx [hunk 8]
```diff
@@ -55,30 +56,51 @@ export default function SettingsPage() {
...
+                <Switch
+                  checked={kernelSmartAccountEnabled}
+                  onChange={SettingsStore.toggleKernelSmartAccountsEnabled}
+                  data-testid="settings-toggle-smart-account-sponsorship"
+                />
+                <Text>{kernelSmartAccountEnabled ? 'Enabled' : 'Disabled'}</Text>
+              </Row>
+
+              <Text h4 css={{ marginBottom: '$5', marginTop: '$5' }}>
+                Sponsorship (Pimlico)
```
Comment (tomiir) [inline advanced/wallets/react-wallet-v2/src/pages/settings.tsx:91]: Should be (`Kernel`) right?

**Review by tomiir — COMMENTED** [review]
For next pr 🤣

File: advanced/wallets/react-wallet-v2/src/hooks/useInitialization.ts (hunk 3 above)
Comment (lukaisailovic) [reply advanced/wallets/react-wallet-v2/src/hooks/useInitialization.ts:36]: Adding safe account atm

File: advanced/wallets/react-wallet-v2/src/lib/KernelSmartAccountLib.ts (hunk 7 above)
Comment (lukaisailovic) [reply advanced/wallets/react-wallet-v2/src/lib/KernelSmartAccountLib.ts:166]: Its an abstraction they made to indicate signer does not have private key and limited capabilities

File: advanced/wallets/react-wallet-v2/src/pages/settings.tsx (hunk 8 above)
Comment (lukaisailovic) [reply advanced/wallets/react-wallet-v2/src/pages/settings.tsx:91]: That currently does nothing as we don't use previous Lib, working on removing it completely and implementing Safe from scratch (with pimlico probably)

**Pull request merged**